### PR TITLE
Emit backward compatible ClusterConfig events

### PR DESCRIPTION
**Pull request by andrejtokarcik**
To ensure backward compatibility after [the `ClusterConfig` split](https://github.com/gravitational/teleport/blob/master/rfd/0028-cluster-config-resources.md):
1. `ClusterConfig` events should also contain those fields that have been moved to a different resource.
2. Updates to the resources that now contain fields previously belonging to `ClusterConfig` should trigger a `ClusterConfig` event in addition to the event of their own kind.

### CODE REVIEW

File: lib/services/local/configuration.go [hunk 1]
```diff
@@ -239,11 +265,11 @@ func (s *ClusterConfigurationService) DeleteClusterConfig() error {
 
 // SetClusterConfig sets services.ClusterConfig on the backend.
 func (s *ClusterConfigurationService) SetClusterConfig(c services.ClusterConfig) error {
-	if err := s.syncClusterConfigWithNetworkingConfig(c); err != nil {
-		return trace.Wrap(err)
+	if c.HasNetworkingFields() {
+		return trace.BadParameter("call SetClusterNetworkingConfig instead")
```
Comment (Joerger) [inline lib/services/local/configuration.go:269]: Error message is a bit unclear
```suggestion
	if c.HasNetworkingFields() {
		return trace.BadParameter("cluster config has legacy networking fields, call SetClusterNetworkingConfig to set these fields")
```

File: lib/services/local/configuration.go [hunk 2]
```diff
@@ -239,11 +265,11 @@ func (s *ClusterConfigurationService) DeleteClusterConfig() error {
...
 	}
-	if err := s.syncClusterConfigWithSessionRecordingConfig(c); err != nil {
-		return trace.Wrap(err)
+	if c.HasSessionRecordingFields() {
+		return trace.BadParameter("call SetSessionRecordingConfig instead")
```
Comment (Joerger) [inline lib/services/local/configuration.go:272]:
```suggestion
	if c.HasSessionRecordingFields() {
		return trace.BadParameter("cluster config has legacy session recording fields, call SetSessionRecordingConfig to set these fields")
```

File: lib/cache/collections.go [hunk 3]
```diff
@@ -1041,10 +1041,15 @@ func (c *clusterConfig) fetch(ctx context.Context) (apply func(ctx context.Conte
 			return nil
 		}
 		c.setTTL(clusterConfig)
+
+		// To ensure backward compatibility, ClusterConfig resources/events may
+		// feature fields that now belong to separate resources/events. Since this
+		// code is able to process the new events, ignore any such legacy fields.
+		// DELETE IN 8.0.0
+		clusterConfig.ClearLegacyFields()
```
Comment (awly) [inline lib/cache/collections.go:1049]: I don't understand why we need to clear the legacy fields at the cache layer
what happens if we keep them?

File: lib/services/local/events.go [hunk 4]
```diff
@@ -157,24 +161,27 @@ func (w *watcher) Error() error {
...
 			if resource == nil {
-				return nil, nil
+				continue
 			}
-			return &services.Event{Type: e.Type, Resource: resource}, nil
+			events = append(events, services.Event{Type: e.Type, Resource: resource})
 		}
 	}
-	return nil, trace.NotFound("no match found for %v %v", e.Type, string(e.Item.Key))
+	return events, errs
```
Comment (awly) [inline lib/services/local/events.go:184]: why is this change required?

File: lib/services/local/events.go [hunk 5]
```diff
@@ -352,32 +365,41 @@ func (p *staticTokensParser) parse(event backend.Event) (services.Resource, erro
...
-			event.Item.Value,
-			services.WithResourceID(event.Item.ID),
-			services.WithExpires(event.Item.Expires),
-			services.SkipValidation(),
-		)
+		// To ensure backward compatibility, do not use the ClusterConfig
+		// resource passed with the event but perform a separate get from the
+		// backend. The resource fetched in this way is populated with all the
+		// fields expected by legacy event consumers.  DELETE IN 8.0.0
+		clusterConfig, err := p.getClusterConfig()
```
Comment (awly) [inline lib/services/local/events.go:402]: this may be breaking the abstraction layers a bit, makes me nervous about cyclical dependencies (in case some code wraps `services.ClusterConfiguration`).
how about extracting `getClusterConfig` into a standalone helper function called from both here and from `GetClusterConfig` to remove a direct dependency between them?

File: lib/services/suite/suite.go [hunk 6]
```diff
@@ -1762,6 +1770,12 @@ func (s *ServicesTestSuite) EventsClusterConfig(c *check.C) {
 				out, err := s.ConfigS.GetClusterConfig()
 				c.Assert(err, check.IsNil)
 
+				// To ensure backward compatibility the ClusterConfig resource is not
+				// emitted in the same form as it is put into the backend, but instead
+				// the event handler performs an additional get of ClusterConfig from
+				// the backend.  Therefore, do not delete the resource immediately but
+				// give some time for the get operation to succeed.  DELETE IN 8.0.0
+				time.Sleep(time.Second)
```
Comment (awly) [inline lib/services/suite/suite.go:1778]: Is there a way to synchronously wait for the event to be emitted here?

**Review by awly — COMMENTED** [review]
To make sure I understand correctly:
- ClusterConfig no longer contains legacy fields when stored in the backend
- when read from the backend, we re-fetch the legacy fields from new resources
- when emitting watch events, we also re-fetch the legacy fields

Is that correct?

File: lib/cache/collections.go (hunk 3 above)
Comment (andrejtokarcik) [reply lib/cache/collections.go:1049]: As you can see this is immediately followed by a `SetClusterConfig` call. If the legacy fields were left intact the following conditions would trigger:
https://github.com/gravitational/teleport/blob/6ce105608f296a797fcb493acf8df32c384e41e7/lib/services/local/configuration.go#L267-L273

File: lib/services/local/events.go (hunk 4 above)
Comment (andrejtokarcik) [reply lib/services/local/events.go:184]: Because a single key can now be matched by multiple parsers each of which produce a different resource/event. For instance, `/cluster_configuration/session_recording` is now matched by both `clusterConfigParser` and `sessionRecordingConfigParser`.

**Comment by andrejtokarcik** [comment]
@awly 
> ClusterConfig no longer contains legacy fields when stored in the backend

That's the ultimate goal but it isn't true yet. As of this PR, `ClusterConfig` still contains the cluster ID, the audit config and some auth related fields: https://github.com/gravitational/teleport/blob/6ce105608f296a797fcb493acf8df32c384e41e7/api/types/types.proto#L512-L538

> when read from the backend, we re-fetch the legacy fields from new resources

Yes, that's covered by `GetClusterConfig`: https://github.com/gravitational/teleport/blob/6ce105608f296a797fcb493acf8df32c384e41e7/lib/services/local/configuration.go#L229-L249

> when emitting watch events, we also re-fetch the legacy fields

Yes, that's why the events service & `clusterConfigParser` now have the `getClusterConfigFunc` dependency: https://github.com/gravitational/teleport/blob/6ce105608f296a797fcb493acf8df32c384e41e7/lib/services/local/events.go#L397-L406

(Normally the resource would be parsed out of the received backend event as in https://github.com/gravitational/teleport/blob/389531307ab7b96c4a6e840d3896bbf4be9d29f9/lib/services/local/events.go#L374-L384 but in this PR we always explicitly call `GetClusterConfig` to trigger the extend-with-other-resources logic implemented there.)

File: lib/services/local/events.go (hunk 5 above)
Comment (andrejtokarcik) [reply lib/services/local/events.go:402]: I don't quite understand the concern nor the suggested solution.

**Review by Joerger — APPROVED** [review]
LGTM overall, though my working knowledge around some of this code is limited.

File: lib/services/local/events.go (hunk 5 above)
Comment (awly) [reply lib/services/local/events.go:397]: `EventsService` already has direct access to the `backend.Backend`.
theoretically, it shouldn't need an opaque function variable to read the full cluster config.
if there was a `func getClusterConfig(bk backend.Backend) (types.ClusterConfig, error)` in this package, you wouldn't need to pass in a callback.

but this is a temporary kludge, I don't mind keeping it for one release cycle.

**Review by awly — APPROVED** [review]

**Review by r0mant — APPROVED** [review]
🤖
